File: src/day03.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define WIDTH 140
#define HEIGHT 140

int is_symbol(const char chr)
{
	return chr != '.' && !isdigit(chr);
}

int any_adjacent(const char grid[HEIGHT][WIDTH], const size_t x1, const size_t x2, const size_t y)
{
	for (size_t x = x1; x <= x2; x++)
	{
		if (y > 0)
		{
			if (x > 0 && is_symbol(grid[y - 1][x - 1]))
			{
				return 1;
			}

			if (x < WIDTH - 1 && is_symbol(grid[y - 1][x + 1]))
			{
				return 1;
			}

			if (is_symbol(grid[y - 1][x]))
			{
				return 1;
			}
		}

		if (y < HEIGHT - 1)
		{
			if (x > 0 && is_symbol(grid[y + 1][x - 1]))
			{
				return 1;
			}

			if (x < WIDTH - 1 && is_symbol(grid[y + 1][x + 1]))
			{
				return 1;
			}

			if (is_symbol(grid[y + 1][x]))
			{
				return 1;
			}
		}
	}

	if (x1 > 0 && is_symbol(grid[y][x1 - 1]))
	{
		return 1;
	}

	if (x2 < WIDTH - 1 && is_symbol(grid[y][x2 + 1]))
	{
		return 1;
	}

	return 0;
}

size_t parse_number(const char grid[HEIGHT][WIDTH], const size_t start, const size_t end, const size_t y)
{
	if (!any_adjacent(grid, start, end, y))
	{
		return 0;
	}

	const size_t len = end - start + 1;
	char buffer[len + 1];
	strncpy(buffer, &grid[y][start], len);
	buffer[len] = '\0';
	return atoi(buffer);
}

int main()
{
	FILE *file = fopen("input/day03", "r");

	int sum1 = 0;
	size_t i = 0;
	char schematic[HEIGHT][WIDTH] = {};

	while (fread(schematic[i++], sizeof(char), WIDTH, file) > 0)
	{
		if (i >= HEIGHT)
		{
			break;
		}

		fread(schematic[i], sizeof(char), 1, file);
	}

	fclose(file);

	for (size_t y = 0; y < HEIGHT; y++)
	{
		size_t start_index = -1;

		for (size_t x = 0; x < WIDTH; x++)
		{
			const char current = schematic[y][x];
			if (start_index == (size_t)-1 && isdigit(current))
			{
				start_index = x;
				continue;
			}

			if (!isdigit(current))
			{
				if (start_index != (size_t)-1)
				{
					sum1 += parse_number(schematic, start_index, x - 1, y);
				}

				start_index = (size_t)-1;
			}
		}

		if (start_index != (size_t)-1)
		{
			sum1 += parse_number(schematic, start_index, WIDTH - 1, y);
		}
	}

	printf("[01] sum: %i\n", sum1);
	return 0;
}

File: day03.lua
---@param str string
---@return boolean
function is_symbol(str)
	return str and str ~= "." and not tonumber(str)
end

---@param grid table
---@param x_start number
---@param x_end number
---@param y number
---@return boolean
function any_adjacent(grid, x_start, x_end, y)
	for x = x_start, x_end do
		if y > 1 then
			if is_symbol(grid[y - 1][x - 1]) then
				return true
			end
			if is_symbol(grid[y - 1][x + 1]) then
				return true
			end
			if is_symbol(grid[y - 1][x]) then
				return true
			end
		end

		if y < #grid then
			if is_symbol(grid[y + 1][x - 1]) then
				return true
			end
			if is_symbol(grid[y + 1][x + 1]) then
				return true
			end
			if is_symbol(grid[y + 1][x]) then
				return true
			end
		end
	end

	if is_symbol(grid[y][x_start - 1]) then
		return true
	end
	if is_symbol(grid[y][x_end + 1]) then
		return true
	end

	return false
end

---@param grid table
---@param x_start number
---@param x_end number
---@param y number
---@return number
function parse_number(grid, x_start, x_end, y)
	if not any_adjacent(grid, x_start, x_end, y) then
		return 0
	end

	local str = ""
	for i = x_start, x_end do
		str = str .. grid[y][i]
	end
	return tonumber(str)
end

local sum1 = 0
local grid = {}

for line in io.lines("input/day03") do
	local row = {}
	for i = 1, #line do
		row[i] = line:sub(i, i)
	end
	table.insert(grid, row)
end

for y = 1, #grid do
	local start_index = 0
	for x = 1, #grid[y] do
		local current = grid[y][x]
		if start_index == 0 and tonumber(current) then
			start_index = x
		elseif not tonumber(current) then
			if start_index > 0 then
				sum1 = sum1 + parse_number(grid, start_index, x - 1, y)
			end
			start_index = 0
		end
	end
	if start_index > 0 then
		sum1 = sum1 + parse_number(grid, start_index, #grid[y], y)
	end
end

print("[01] sum:", sum1)